src/serde/tree_cache.rs

```diff
@@ -0,0 +1,744 @@
+use crate::serde::serialized_length_atom;
+use crate::serde::RandomState;
+use crate::serde::VisitedNodes;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+const MIN_SERIALIZED_LENGTH: u64 = 4;
+
+#[derive(Clone, Debug)]
+struct NodeEntry {
+    /// the tree hash of this node. It may be None if it or any of its children
+    /// is the sentinel node, which means we can't compute the tree hash.
+    tree_hash: Option<Bytes32>,
+    /// a node can have an arbitrary number of parents, since they can be reused
+    /// this is a list of parent nodes, followed by whether we're the left or
+    /// right child. The u32 is an index into the node_entry vector.
+    parents: Vec<(u32, ChildPos)>,
+    /// if this node doesn't have a tree_hash, the serialized length is not
+    /// valid as it cannot be computed.
+    serialized_length: u64,
+    /// set to non-zero, if this node is pushed onto the parse stack. Since a
+    /// node can be on the stack multiple times, it's counted up every time it's
+    /// pushed, and counted down every time it's popped.
+    pub on_stack: u32,
+}
+
+struct PartialPath {
+    // the path we've built so far
+    path: PathBuilder,
+    // if we're traversing the stack, this is the stack position. Note that this
+    // is not an index into the stack array, it's a counter of how far away from
+    // the top of the stack we are. 0 means we're at the top, and we've found
+    // a path.
+    // -1 means we're not traversing the stack.
+    // TODO: these fields are mutually exclusive (stack_pos vs. idx and child).
+    // it might be nice to use an enum, but it's not obvious that it's simpler
+    // or faster
+    stack_pos: i32,
+    // if we're not traversing the stack, this is the next parent, and whether
+    // we're coming from the left or right child
+    idx: u32,
+    child: ChildPos,
+}
+
+enum CacheOp {
+    // node to traverse
+    Traverse(NodePtr),
+    // pair node
+    Cons(NodePtr),
+}
+
+#[derive(Clone)]
+pub struct TreeUndoState {
+    stack: Vec<u32>,
+    lookup: VisitedNodes,
+    sentinel_entry: Option<u32>,
+}
+
+#[derive(Default)]
+pub struct TreeCache {
+    /// caches extra metadata about a tree of nodes. The value is an index into
+    /// the node_entry vector.
+    cache: HashMap<NodePtr, u32>,
+
+    /// The metadata for all nodes in the tree. This is like a shadow tree
+    /// structure to the NodePtr one. The most important difference is that
+    /// identical nodes are merged, using the same NodeEntry, and additional
+    /// metadata is kept, such as the tree hash.
+    node_entry: Vec<NodeEntry>,
+
+    /// maps tree-hashes to the index of the corresponding NodeEntry in the
+    /// node_entry vector. For any given tree hash, we're only supposed to
+    /// have a single NodeEntry. There may be multiple NodePtr referring to
+    /// the same NodeEntry (if they are identical sub trees).
+    hash_to_node: HashMap<Bytes32, u32, RandomState>,
```

Comment:
Similarly, `hash_to_node_entry_index_map` (or do `s/node_entry/nodes/g` and then use `hash_to_node_index`). I'd argue that it's not a hash-to-node because what comes out is a u32, which is not a node.

Reply:
I implemented your suggested optimization in a separate patch. It gives 10-25% performance improvement, which is quite significant. I'll post that PR separately, but there I'm renaming this field.